src/context.ts

```diff
@@ -464,17 +464,33 @@ export class Context<
    */
   set res(_res: Response | undefined) {
     this.#isFresh = false
-    if (this.#res && _res) {
-      this.#res.headers.delete('content-type')
-      for (const [k, v] of this.#res.headers.entries()) {
-        if (k === 'set-cookie') {
-          const cookies = this.#res.headers.getSetCookie()
-          _res.headers.delete('set-cookie')
-          for (const cookie of cookies) {
-            _res.headers.append('set-cookie', cookie)
+    if (_res) {
+      try {
+        // Check if `_res` headers are immutable.
+        _res.headers.delete('______________________________')
+      } catch (e) {
+        if (e instanceof TypeError && e.message.includes('immutable')) {
+          // `_res` is immutable (probably a response from a fetch API), so retry with a new response.
+          this.res = new Response(_res.body, {
+            headers: _res.headers,
+            status: _res.status,
+          })
+          return
+        }
+      }
+
+      if (this.#res) {
+        for (const [k, v] of this.#res.headers.entries()) {
+          if (k === 'set-cookie') {
+            const cookies = this.#res.headers.getSetCookie()
+            _res.headers.delete('set-cookie')
+            for (const cookie of cookies) {
+              _res.headers.append('set-cookie', cookie)
+            }
+          } else if (!_res.headers.has(k)) {
```

Comment:
Why wasn't this checked before?

Reply:
I think that point is correct. I think the new response headers should be prioritised. And you're right, it would be a spec change and should be discussed in another PR.